### ISSUE TITLE
[java] UnusedPrivateMethod false positives with lombok.val

### ISSUE THREAD

**Issue by pankratz76** [pmd/pmd#5369]
**Affects PMD Version:** 7.10.0

**Rule:** [UnusedPrivateMethod](https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#unusedprivatemethod)

**Description:**
When using [Lombok val](https://projectlombok.org/features/val), the type of the local variable is `lombok.val` instead of the inferred correct type.
When such a variable is used in a method call, we don't find the correct method declaration due to the different types and report the called method as unused.

Workaround: Don't use `val`, use `final var` instead. Note: lombok val is a _final_ local variable, while Java's var, is a non-final local variable.

**Code Sample demonstrating the issue:**

```java
package sample;

import lombok.val;

class Foo {
  public void method() {
     val test = "a";

     privateMethod(test);
  }

  private void privateMethod(String a) {
    System.out.println(a);
  }
}
```

Complete reproducer project: [issue-5369-unusedprivatemethod.zip](https://github.com/user-attachments/files/18904657/issue-5369-unusedprivatemethod.zip)

**Expected outcome:**

PMD reports a violation at line 12, but that's wrong. That's a false positive.

**Running PMD through:** Maven

**Comment by oowekyala** [comment]
This is likely a problem with your plugin configuration. PMD needs compiled classes to be on the classpath in order to resolve external method calls. If the code is as simple as you show, it should have no problem as it only uses methods in the current class. But if it uses methods from somewhere else then their type is unknown unless the classpath is populated. Please share what plugin you are using and your configuration for it (eg maven pom). Ideally make a minimal reproducible example in the form of a small repository. Thank you.

**Comment by AlexisJehan** [comment]
I have the same issue in two projects.

Here is a reproductible example:
[issue-5369.zip](https://github.com/user-attachments/files/18274545/issue-5369.zip)
(Issue with two methods have the same name in the same class)

**Comment by punkratz312** [comment]
@AlexisJehan Thanks for the demo! I've tried merging it into the test, and it seems to work as expected. Could you maybe reproduce this?

[GitHub Pull Request #5470](https://github.com/pmd/pmd/pull/5470)

**Comment by adangel** [comment]
> However, when the rule is applied via the plugin in the full project, PMD incorrectly reports the private methods as unused. Despite this, the tests show that the methods are indeed used, as commenting them out causes test failures, confirming their necessity.

@punkratz312 Please tell us, _how_ you call PMD... What is "the plugin"? thanks

**Comment by punkratz312** [comment]
We call the plugin via mvn within our build phase: 

`[ERROR] Failed to execute goal org.apache.maven.plugins:maven-pmd-plugin:3.26.0:check (default) on project tm-core-srv: PMD 7.7.0 has found 78 violations.`

![Image](https://github.com/user-attachments/assets/8a0dad40-faf6-4d23-8ada-435d0be8553f)

**Comment by adangel** [comment]
@punkratz312 You code sample with the class `NullScript` does _not_ show an unused method. PMD can even analyse this file without having it compiled before, as it doesn't use any other types. I guess, this example is too simple. Can you provide another example?

Also, can you share your pom.xml file? We need to double check, that maven-pmd-plugin is used correctly.


@AlexisJehan Thanks for your example. However, you have a mistake in your maven configuration: In your configuration, maven-pmd-plugin is executed in phase "generate-sources" which is too early. In that phase, the java files have not been compiled. Just remove the line 28 (`<phase>generate-sources</phase>`) and maven-pmd-plugin will run after compilation (it will run in phase "verify"). That means, just run `mvn verify` and everything works.

**Comment by punkratz312** [comment]
nothing special:
```
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-pmd-plugin</artifactId>
                <version>3.26.0</version>
                <configuration>
                    <analysisCache>true</analysisCache>
                    <analysisCacheLocation>${project.basedir}/.pmd/.cache</analysisCacheLocation>
                    <excludeFromFailureFile>${project.basedir}/.pmd/exclude.properties</excludeFromFailureFile>
                    <printFailingErrors>true</printFailingErrors>
                    <rulesets>
                        <ruleset>${project.basedir}/.pmd/ruleset.xml</ruleset>
                    </rulesets>
                </configuration>
                <executions>
                    <execution>
                        <goals>
                            <goal>check</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```

```
<ruleset xmlns="http://pmd.sourceforge.net/ruleset/2.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" name="Default Maven PMD Plugin Ruleset"
         xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 http://pmd.sourceforge.net/ruleset_2_0_0.xsd">
    <description>tmp-fix</description>
    <rule ref="category/java/bestpractices.xml"><!--TODO revise to apply best practices-->
        <exclude name="GuardLogStatement"/><!--logging bug: -->
        <exclude name="LiteralsFirstInComparisons"/>
        <exclude name="UnusedPrivateMethod"/><!--false positive-->
        <exclude name="UseVarargs"/><!--intend-->
    </rule>
</ruleset>
```

**Comment by AlexisJehan** [comment]
> your maven configuration: In your configuration, maven-pmd-plugin is executed in phase "generate-sources" which is too early. In that phase, the java files have not been com

Effectively, thank you I was not aware about this, this fix my issue.

**Comment by oowekyala** [comment]
I think the easiest thing would be to implement part of #3119. We could just infer the variable type when it has type `lombok.val` or `lombok.var`, it's a very small change I think. Probably the easiest route is to make `ASTVariableId::isTypeInferred` return true for those variable types. We could add a language property to disable this behavior, although it should probably be enabled by default.

This has the advantage that all rules can benefit from it, and we would get precise overload resolution instead of one workaround in each rule.

For #5079, something similar could be done. We could make ASTVariableId::isFinal care about lombok.val.
